### PR TITLE
[FEATURE] - Standup Feature Update(Milestone-2)

### LINKED ISSUE #847: [FEATURE] - Standup Feature Update(Milestone-2)
### Parent Issue
- #837 

### Issues
* Add one input box after the other input box to add standup.

### PR DESCRIPTION
**Pull request by Pratiyushkumar**
### Issue:
**Issue Ticket**
- Closes #847 

**Parent Issue -**
- #837 
### Description:
Overview
As the Standup feature is deployed on Production under the feature flag, remove the feature from the feature flag I asked developers in the team for review and I got a few suggestions from developers. so this project will be an updated version of the existing feature in production.

Context
As per the suggestions received, this project is important from a user perspective because this project will make the user experience much better and also will remove the confusion between the users while filling up the standup form.

Issues to Be Addressed - 
 * [x] Dynamically input field should be added using the "+"  button
 * [x]  Form should be submitted
 * [x]  Testing should be done

### Dev Tested:
- [x] Yes

### Test Coverage
image
image

###Video Recording

https://github.com/Real-Dev-Squad/website-status/assets/82165483/cffc8910-4df8-40f4-aba0-8168fa96ac3c

### CODE REVIEW

**Comment by vinit717** [comment]
In the video you shared progresses API is failing what is the case of it failing

File: __tests__/Unit/Components/Standup/StandupContainer.test.tsx [hunk 1]
```diff
@@ -2,266 +2,23 @@ import StandUpContainer from '@/components/standup';
...
-        await waitFor(() => {
-            expect(screen.getByText(ERROR_MESSAGE)).toBeInTheDocument();
-        });
+        screen.debug();
+        const submitButton = screen.getByRole('button', { name: 'Submit' });
+        const addButton = container.getElementsByClassName('addButton');
+        expect(screen.getByText('Standup Update')).toBeInTheDocument();
+        expect(screen.getByText('Yesterday')).toBeInTheDocument();
+        expect(screen.getByText('Today')).toBeInTheDocument();
+        expect(screen.getByText('Blocker')).toBeInTheDocument();
```
[the remark this answers is not in the document]
Comment (Pratiyushkumar) [reply __tests__/Unit/Components/Standup/StandupContainer.test.tsx:20]: I didn't get your this comment, can you please elaborate

File: src/components/standup/FormInputComponent.tsx [hunk 2]
```diff
@@ -1,36 +1,131 @@
...
-                required
-                name={name}
-                value={value}
-                onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
-                    handleChange(e)
+const intialSection = [
+    { title: 'Yesterday', inputs: [''] },
+    { title: 'Today', inputs: [''] },
+    { title: 'Blocker', inputs: [''] },
+];
```
[the remark this answers is not in the document]
Comment (Pratiyushkumar) [reply src/components/standup/FormInputComponent.tsx:11]: your suggestion is correct and I tried this but when I am submitting the form the form is not getting reset by passing initalSection in setSections(), When I pass the array of objects the form is getting updated.

[the remark this answers is not in the document]
Comment (Pratiyushkumar) [reply src/components/standup/FormInputComponent.tsx:11]: if you have any solution then please let me know

**Comment by Pratiyushkumar** [comment]
> In the video you shared progresses API is failing what is the case of it failing

The Get API of /progressess is getting called twice after form submission, once it is failing and again it is running...I tried finding solution but did not get it. request you to help me out.

File: __tests__/Unit/Components/Standup/StandupContainer.test.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (sahsisunny) [reply __tests__/Unit/Components/Standup/StandupContainer.test.tsx:20]: What specific input element are you testing? Do you want to test any element, such as text, textarea, etc.? Please specify what you want to test.

File: src/components/standup/FormInputComponent.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (sahsisunny) [reply src/components/standup/FormInputComponent.tsx:11]: Can you share the screenshot of the error when you are using the initialState?

[the remark this answers is not in the document]
Comment (Pratiyushkumar) [reply src/components/standup/FormInputComponent.tsx:11]: I am not getting any error, it's just that after submitting the form fields should get empty which is not getting empty

File: __tests__/Unit/Components/Standup/StandupContainer.test.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (Pratiyushkumar) [reply __tests__/Unit/Components/Standup/StandupContainer.test.tsx:20]: basically In the current component that I am testing, that component renders a dom which is that initial look of the form, in this test I am testing that initial look after rendering for first time what will happen, what are the things I am able to get.

**Comment by Pratiyushkumar** [comment]
> > In the video you shared progresses API is failing what is the case of it failing
> 
> The Get API of /progressess is getting called twice after form submission, once it is failing and again it is running...I tried finding solution but did not get it. request you to help me out.

done @vinit717, thank you for the help

File: src/components/standup/FormInputComponent.tsx [hunk 3]
```diff
@@ -1,36 +1,161 @@
...
+                </div>
+            ))}
+            <button
+                type="submit"
+                disabled={!isFormValid()}
+                className={`${styles.submitButton}`}
+            >
+                Submit
+            </button>
+        </form>
```
Comment (vinit717) [inline src/components/standup/FormInputComponent.tsx:158]: Can you break down this

Comment (Pratiyushkumar) [reply src/components/standup/FormInputComponent.tsx:158]: any solution you would like to suggest me?

Comment (vinit717) [reply src/components/standup/FormInputComponent.tsx:158]: import React, { FC, useState } from 'react';
import { useRouter } from 'next/router';
import { useSaveProgressMutation } from '@/app/services/progressesApi';
import { toast, ToastTypes } from '@/helperFunctions/toast';
import StandupSection from './StandupSection';
import styles from '@/components/standup/standupContainer.module.scss';

const initialSections = [
  { title: 'Yesterday', inputs: [''] },
  { title: 'Today', inputs: [''] },
  { title: 'Blocker', inputs: [''] },
];

const FormInputComponent: FC = () => {
  const router = useRouter();
  const [sections, setSections] = useState(initialSections);
  const [addStandup] = useSaveProgressMutation();
  const { SUCCESS, ERROR } = ToastTypes;

  const handleAddField = (sectionIndex: number) => {
    const newSections = [...sections];
    newSections[sectionIndex].inputs.push('');
    setSections(newSections);
  };

  const handleRemoveField = (sectionIndex: number) => {
    const newSections = [...sections];
    newSections[sectionIndex].inputs.pop();
    setSections(newSections);
  };

  const handleInputChange = (
    sectionIndex: number,
    inputIndex: number,
    value: string
  ) => {
    const newSections = [...sections];
    newSections[sectionIndex].inputs[inputIndex] = value;
    setSections(newSections);
  };

  const isFormValid = () => {
    return sections.every((section) =>
      section.inputs.every((input) => input.trim() !== '')
    );
  };

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const newData = {
      type: 'user',
      completed: sections[0].inputs.join('. '),
      planned: sections[1].inputs.join('. '),
      blockers: sections[2].inputs.join('. '),
    };
    try {
      const data = await addStandup(newData).unwrap();
      toast(SUCCESS, data.message);
    } catch (error) {
      toast(ERROR, error.data.message);
    }
    setSections(initialSections);
    router.replace(router.asPath);
  };

  return (
    <form className={styles.standupForm} onSubmit={handleSubmit} aria-label="form">
      {sections.map((section, sectionIndex) => (
        <StandupSection
          key={sectionIndex}
          section={section}
          sectionIndex={sectionIndex}
          onInputChange={handleInputChange}
          onAddField={handleAddField}
          onRemoveField={handleRemoveField}
        />
      ))}
      <button type="submit" disabled={!isFormValid()} className={styles.submitButton}>
        Submit
      </button>
    </form>
  );
};

export default FormInputComponent;

import React from 'react';
import StandupInput from './StandupInput';

const StandupSection = ({
  section,
  sectionIndex,
  onInputChange,
  onAddField,
  onRemoveField,
}) => {
  return (
    <div className={styles.formFields}>
      <label className={styles.updateHeading} htmlFor={section.title}>
        {section.title}
      </label>
      <div className={styles.inputContainers}>
        <div className={styles.inputBoxContainer}>
          {section.inputs.map((inputValue, inputIndex) => (
            <StandupInput
              key={inputIndex}
              sectionIndex={sectionIndex}
              inputIndex={inputIndex}
              inputValue={inputValue}
              onInputChange={onInputChange}
              onAddField={onAddField}
              onRemoveField={onRemoveField}
              sectionTitle={section.title}
            />
          ))}
        </div>
      </div>
    </div>
  );
};

export default StandupSection;


import React from 'react';

const StandupInput = ({
  sectionIndex,
  inputIndex,
  inputValue,
  onInputChange,
  onAddField,
  onRemoveField,
  sectionTitle,
}) => {
  return (
    <div className={styles.inputAndButtons}>
      <div className={styles.inputBox}>
        <input
          className={styles.inputField}
          data-testid={`${sectionTitle}${inputIndex}`}
          name={sectionTitle}
          type="text"
          required
          value={inputValue}
          onChange={(e) => onInputChange(sectionIndex, inputIndex, e.target.value)}
        />
      </div>
      <div className={styles.buttonsContainer}>
        {inputIndex === section.inputs.length - 1 && (
          <button className={styles.addButton} type="button" onClick={() => onAddField(sectionIndex)}>
            +
          </button>
        )}
        {inputIndex === section.inputs.length - 1 && section.inputs.length > 1 && (
          <button className={styles.removeButton} type="button" onClick={() => onRemoveField(sectionIndex)}>
            x
          </button>
        )}
      </div>
    </div>
  );
};

export default StandupInput;

**Review by vinit717 — COMMENTED** [review]
Resolve all the things we discussed over call in the next pr

**Review by vinit717 — APPROVED** [review]

**Review by sahsisunny — APPROVED** [review]

**Pull request merged**